docs/design_proposals/debug-events.md

```diff
@@ -0,0 +1,187 @@
+duplicate functionality already being performed by Skaffold, and also requiring this functionality to be
+re-implemented in each IDE.
+
+This design proposal recommends that Skaffold instead also watch and issue events when debuggable
+containers are started and terminated.
+
+----
+## Design
+
+This design proposes that Skaffold issue events to inform the IDEs of the following conditions:
+
+- The appearance of a container that is debuggable.  This event will provide:
+    - the namespace.
+    - the pod ID
+    - the container ID
+    - the contents of the `debug.cloud.google.com/config` configuration block for the container
+      (a JSON object) including (! indicates new items)
+         - the language runtime type (`go`, `jvm`, `nodejs`, `python`)
+         - ! the image name (which should allow cross-referencing with the `skaffold.yaml`)
+         - the port(s)
+         - ! the container image's working directory
+         - ! (optional) the container image's _remote root_ for source file resolving 
+
+- The disappearance of a container that was debuggable.  This event will provide:
+    - the namespace
+    - the pod ID
+    - the container ID
+    - the contents of the `debug.cloud.google.com/config` configuration block for the container
+      (a JSON object) including (! indicates new items)
+         - the language runtime type (`go`, `jvm`, `nodejs`, `python`)
+         - ! the corresponding artifact's `name` (which should allow cross-referencing with the `skaffold.yaml`)
+         - the port(s)
+         - ! the container image's working directory
+         - ! (optional) the container image's _remote root_ for source file resolving 
+         
+Skaffold will also maintain state and report on the set of debuggable containers.
+
+With this information, the IDE should be able to:
+
+- Identify the corresponding artifact within the `skaffold.yaml`
+- Identify the corresponding artifact from a port-forward event by correlating the
+  remote port with the debug port from the previously-raised debug event
+- Establish new debug launch configurations with the forwarded port,
+  and the local source location, and any other necessary information
+- Tear down existing debug launch configurations when a container is terminated
+- (*In the future*, *if necessary*) Request port-forward of the debug port for the
+  namespace/pod/container/port on-demand
+ 
+### Example
+
+```json
+{
+  "result": {
+    "timestamp": "2019-10-23T18:50:38.916931Z",
+    "event": {
+      "deployEvent": {
+        "status": "In Progress"
+      }
+    },
+    "entry": "Deploy started"
+  }
+}
+{
+  "result": {
+    "timestamp": "2019-10-23T18:50:39.629826Z",
+    "event": {
+      "deployEvent": {
+        "status": "Complete"
+      }
+    },
+    "entry": "Deploy complete"
+  }
+}
+{
+  "result": {
+    "timestamp": "2019-10-23T18:50:41.004396Z",
+    "event": {
+      "debugEvent": {
+        "status": "Started",
+        "podName": "npm",
```

Comment:
What is the best way to map a `debugEvent` to a `portEvent`? Is it by comparing `podName` and `containerName`?

Reply:
And namespace.

Reply:
The debug-events will be raised before the port-events since debug-events will be raised when the container is started, and that's when the port-forward will be initiated.